algorithms/src/snark/marlin/ahp/prover/state.rs

```diff
@@ -70,95 +52,142 @@ pub struct State<'a, F: PrimeField, MM: MarlinMode> {
+                    lhs_polynomials: None,
+                };
+                Ok((circuit, state))
             })
-            .collect();
-        let batch_size = private_variables.len();
-        assert_eq!(padded_public_input.len(), batch_size);
+            .collect::<SynthesisResult<BTreeMap<_, _>>>()?;
+
+        let max_constraint_domain = max_constraint_domain.ok_or(AHPError::BatchSizeIsZero)?;
+        let max_non_zero_domain = max_non_zero_domain.ok_or(AHPError::BatchSizeIsZero)?;
 
         Ok(Self {
-            index,
-            input_domain,
-            constraint_domain,
-            non_zero_a_domain,
-            non_zero_b_domain,
-            non_zero_c_domain,
-            batch_size,
-            padded_public_variables: padded_public_input,
-            x_poly,
-            private_variables,
-            z_a: None,
-            z_b: None,
+            max_constraint_domain,
+            max_non_zero_domain,
+            circuit_specific_states,
+            total_instances,
             first_round_oracles: None,
-            mz_poly_randomizer: None,
             verifier_first_message: None,
-            lhs_polynomials: None,
-            sums: None,
         })
     }
 
-    /// Get the batch size.
-    pub fn batch_size(&self) -> usize {
-        self.batch_size
+    /// Get the batch size for a given circuit.
+    pub fn batch_size(&self, circuit: &Circuit<F, MM>) -> Option<usize> {
+        self.circuit_specific_states.get(circuit).map(|s| s.batch_size)
     }
 
     /// Get the public inputs for the entire batch.
-    pub fn public_inputs(&self) -> Vec<Vec<F>> {
-        self.padded_public_variables.iter().map(|v| super::ConstraintSystem::unformat_public_input(v)).collect()
+    pub fn public_inputs(&self, circuit: &Circuit<F, MM>) -> Option<Vec<Vec<F>>> {
+        // We need to export inputs as they live longer than prover_state
+        self.circuit_specific_states.get(circuit).map(|s| {
+            s.padded_public_variables.iter().map(|v| super::ConstraintSystem::unformat_public_input(v)).collect()
+        })
     }
 
     /// Get the padded public inputs for the entire batch.
-    pub fn padded_public_inputs(&self) -> Vec<Vec<F>> {
-        self.padded_public_variables.clone()
-    }
-
-    pub fn fft_precomputation(&self) -> &FFTPrecomputation<F> {
-        &self.index.fft_precomputation
+    pub fn padded_public_inputs(&self, circuit: &Circuit<F, MM>) -> Option<&[Vec<F>]> {
+        self.circuit_specific_states.get(circuit).map(|s| s.padded_public_variables.as_slice())
     }
 
-    pub fn ifft_precomputation(&self) -> &IFFTPrecomputation<F> {
-        &self.index.ifft_precomputation
+    /// Iterate over the lhs_polynomials
+    pub fn lhs_polys_into_iter(self) -> impl Iterator<Item = DensePolynomial<F>> {
+        self.circuit_specific_states
+            .into_values()
+            .flat_map(|s| s.lhs_polynomials.unwrap().into_iter())
+            .collect::<Vec<_>>()
+            .into_iter()
```

Comment:
```suggestion
```
aren't these redundant?

Reply:
This looks not as straightforward:
```
hidden type for `impl Iterator<Item = dense::DensePolynomial<F>>` captures lifetime that does not appear in boundsrustcClick for full compiler diagnostic
state.rs(98, 6): hidden type `std::iter::FlatMap<std::collections::btree_map::IntoValues<&'a circuit::Circuit<F, MM>, snark::marlin::ahp::prover::state::CircuitSpecificState<F>>, std::array::IntoIter<dense::DensePolynomial<F>, 3>, [closure@algorithms/src/snark/marlin/ahp/prover/state.rs:189:23: 189:26]>` captures the lifetime `'a` as defined here
```

Reply:
isn't the missing lifetime `'a`? i.e. `-> impl Iterator<Item = DensePolynomial<F>> + 'a`?

Reply:
Brilliant, TIL, ca149616c